Add Multi-tenancy support for Reactive Resource Server

rhamedy opened the pull request:
Fixes: gh-6727
Fixes: gh-6723

rhamedy commented:
@jzheaux I have decided to create a draft pull request to get some feedback before proceeding with updating the `javadoc` and adding tests. 

Seems like draft pull requests does not trigger build, I wish it did that way we would have figured whether the changes added break any tests or not.

jzheaux commented on web/src/main/java/org/springframework/security/web/server/authentication/AuthenticationWebFilter.java:
```diff
@@ -80,7 +81,16 @@
+
+	/**
+	 * Creates an instance
+	 * @param authenticationManager the authentication manager to use
```
Please add `@since 5.2` here.

rhamedy commented on core/src/main/java/org/springframework/security/authentication/ReactiveAuthenticationManagerResolver.java:
```diff
@@ -0,0 +1,8 @@
+
+@FunctionalInterface
+public interface ReactiveAuthenticationManagerResolver<C> {
+	ReactiveAuthenticationManager resolve(C context);
```
That make sense. I initially had `Mono<..>` but, then changed it probably because a part of code the code failed for some reason but, now that we are reducing the change to `Resource Server` only this should be fine 👍

rhamedy commented on web/src/main/java/org/springframework/security/web/server/authentication/AuthenticationWebFilter.java:
```diff
@@ -80,7 +81,16 @@
+	 * Creates an instance
+	 * @param authenticationManager the authentication manager to use
+	 */
+	public AuthenticationWebFilter(ReactiveAuthenticationManagerResolver<ServerHttpRequest> reactiveAuthenticationManagerResolver) {
```
That make sense! I got carried away by word `reactive` 😄

rhamedy commented on config/src/main/java/org/springframework/security/config/web/server/ServerHttpSecurity.java:
```diff
@@ -1322,8 +1347,12 @@ public OAuth2ResourceServerSpec and() {
+
+				this.reactiveAuthenticationManagerResolver = this.reactiveAuthenticationManagerResolver != null
+						? this.reactiveAuthenticationManagerResolver
+						: request -> getAuthenticationManager();
```
Curious how does the above suggested change do what you said? Isn't everything inside `configure` method executed at `configuration` time? 🤔

jzheaux replied:
Good question. If we did

```java
this.reactiveAuthenticationManagerResolver = request -> getAuthenticationManager()
```

then `getAuthenticationManager()` wouldn't get invoked until an HTTP request came in and `resolver.resolve` was invoked.

Another way to look at this is remembering that a lambda is like a compressed inner class:

```java
this.reactiveAuthenticationManagerResolver = new ReactiveAuthenticationManagerResolver<>() {
    public AuthenticationManager resolve(HttpServerRequest request) {
        return getAuthenticationManager();
    }
}
```

rhamedy replied:
Beautiful! 😍

rhamedy commented on config/src/main/java/org/springframework/security/config/web/server/ServerHttpSecurity.java:
```diff
@@ -1153,6 +1164,13 @@ public OAuth2ResourceServerSpec bearerTokenConverter(ServerAuthenticationConvert
 			return this;
 		}
 
+		public OAuth2ResourceServerSpec reactiveAuthenticationManagerResolver(
```
I have updated the names across the PR to keep them consistent 👍

rhamedy commented:
@jzheaux thank you for the initial review. I think `draft` was a good idea as I wasn't sure if I have understood the scope of work or not. I have addressed most of your comments and pushed my changes up for review. 
I have not written test for all cases and changes, I only added 1 test for the `AuthenticationWebFilterTests` class for now. 
I am curious to what we can and cannot test with regards to changes in this PR?

jzheaux commented on web/src/main/java/org/springframework/security/web/server/authentication/AuthenticationWebFilter.java:
```diff
@@ -95,7 +112,12 @@ public AuthenticationWebFilter(ReactiveAuthenticationManager authenticationManag
+
+		ReactiveAuthenticationManager authenticationManager = this.authenticationManagerResolver
+				.resolve(exchange.getRequest())
+				.block();
```
We actually can't introduce blocking operations into our reactive code. Instead, you'd want to map the result of this:

```java
return this.authenticationManagerResolver.resolve(exchange.getRequest())
		.map(authenticationManager -> authenticationManager.authenticate(token))
		.switchIfEmpty(...
```

rhamedy replied:
I had to cast `Mono<Authentication>` to `Authentication` to be able to use existing `onAuthenticationSuccess` as follow  

```
.flatMap(authentication -> onAuthenticationSuccess((Authentication) authentication, webFilterExchange))
``` 
Any thoughts?

rhamedy replied:
From the look of `test` failures, turns out I cannot do above 🤔

rhamedy replied:
At the moment, the only way that I can get the `flatMap` working is with `block` using existing `onAuthenticationSuccess` as follow 

```
.flatMap(authentication -> onAuthenticationSuccess(authentication.block(), webFilterExchange))
``` 
And since we should not use `blocking` operation in `reactive` then we might have to write a `onAuthenticationSuccess` that takes `Mono<Authentication>` or if there is a another way? 🤔

jzheaux commented on config/src/main/java/org/springframework/security/config/web/server/ServerHttpSecurity.java:
```diff
@@ -1266,7 +1285,19 @@ public void setBearerTokenConverter(ServerAuthenticationConverter bearerTokenCon
+			 * @param authenticationManagerResolver the authentication manager resolver
+			 * @return the {@code JwtSpec} for additional configuration
+			 */
+			public JwtSpec authenticationManager(
```
This method shouldn't be necessary. The user can supply an `authenticationManagerResolver` at the `oauth2` level or an `authenticationManager` at the `jwt` or `opaqueToken` levels.

jzheaux commented on web/src/main/java/org/springframework/security/web/server/authentication/AuthenticationWebFilter.java:
```diff
@@ -95,11 +112,13 @@ public AuthenticationWebFilter(ReactiveAuthenticationManager authenticationManag
-		return this.authenticationManager.authenticate(token)
+
+		return this.authenticationManagerResolver.resolve(exchange.getRequest())
+			.map(authenticationManager -> authenticationManager.authenticate(token))
```
@rhamedy I think what's necessary here is:

```suggestion
			.flatMap(authenticationManager -> authenticationManager.authenticate(token))
```

It works like `Optiona#flatMap` does, in that if a mapping method returns an `Optional`, then `flatMap` will strip it.

You should also be able to remove the cast at that point from line 119.

rhamedy replied:
That make sense! 👍

rhamedy commented:
Hi @jzheaux 

The switch to `flatMap` was what I needed. Pushed my changes as well as added some unit tests to the `AuthenticationWebFilter`. 
Not sure if we have to write tests for changes in `OAuth2ResourceServerSpec`? If yes, where can I find existing tests? The `ServerHttpSecurityTest` does not seem to have any for `OAuth2ResourceServerSpec` 😕 

Also, please let me know if any changes need to be made in method names (naming is not my strongest skill set), javadoc, and remaining tests (if any). 

Thanks for your feedback.

jzheaux commented on config/src/main/java/org/springframework/security/config/web/server/ServerHttpSecurity.java:
```diff
@@ -1254,7 +1273,7 @@ public void setBearerTokenConverter(ServerAuthenticationConverter bearerTokenCon
 		 */
 		public class JwtSpec {
-			private ReactiveAuthenticationManager authenticationManager;
+			private ReactiveAuthenticationManagerResolver<ServerHttpRequest> authenticationManagerResolver;
```
The tests for these ought to go into [OAuth2ResourceServerSpecTests](https://github.com/spring-projects/spring-security/blob/master/config/src/test/java/org/springframework/security/config/web/server/OAuth2ResourceServerSpecTests.java).

rhamedy replied:
@jzheaux struggling with setting up the `configuration` for this one. I have played around with variation of the following and the results I get are 

- No qualifying bean for `ReactiveJwtDecoder` and when I declare that when stubs then it results in status 200 ok ignoring the authentication manager and resolver 😕 
- Without `ReactiveJwtDecoder` I get `500 error` 

```
@EnableWebFlux
@EnableWebFluxSecurity
static class CustomAuthenticationManagerResolverConfig {
	@Bean
	SecurityWebFilterChain springSecurity(ServerHttpSecurity http) throws Exception {
		// @formatter:off
		http
			.authorizeExchange()
				.pathMatchers("/**/message/**").hasAnyAuthority("SCOPE_message:read")
				.and()
			.oauth2ResourceServer()
				.authenticationManagerResolver(authenticationManagerResolver())
				.jwt();
		// @formatter:on
		return http.build();
	}

	@Bean
	ReactiveAuthenticationManagerResolver<ServerHttpRequest> authenticationManagerResolver() {
		return mock(ReactiveAuthenticationManagerResolver.class);
	}
	@Bean
	ReactiveAuthenticationManager authenticationManager() {
		return mock(ReactiveAuthenticationManager.class);
	}
}
```
if I chain the `authenticationManager` after `jwt()` as follow 

```
.oauth2ResourceServer()
		.authenticationManagerResolver(authenticationManagerResolver())
	        .jwt()
		.authenticationManager(authenticationManager())
``` 
then it works but, then how is that a test of `authenticationManagerResolver` then? 🤔

jzheaux replied:
So, you wouldn't supply both `jwt()` and `authenticationManagerResolver()` together. (There is a separate ticket to clean this up).

If you do:

```java
@EnableWebFlux
@EnableWebFluxSecurity
static class CustomAuthenticationManagerResolverConfig {
	@Bean
	SecurityWebFilterChain springSecurity(ServerHttpSecurity http) throws Exception {
		// @formatter:off
		http
			.authorizeExchange()
				.pathMatchers("/**/message/**").hasAnyAuthority("SCOPE_message:read")
				.and()
			.oauth2ResourceServer()
				.authenticationManagerResolver(authenticationManagerResolver());
		// @formatter:on
		return http.build();
	}

	@Bean
	ReactiveAuthenticationManagerResolver<ServerHttpRequest> authenticationManagerResolver() {
		return mock(ReactiveAuthenticationManagerResolver.class);
	}
}
```

then does it work?

rhamedy replied:
I am certain that I have given that a try and came across an issue. I have tried a few variations of the configurations and got errors messages such as 
- A `404` error
- An error complaining about `ReactiveJwtDecoder` not being there, although I feel like this is due to use of `jwt()` 
- A `500` null pointer exception in some `reactive` lamda 

I can give it another try later in the day and push my tests as well as share the error messages! :slightly_smiling_face:

rhamedy commented:
Hey @jzheaux sorry, I had some `git hiccup` and had to do a few `hard resets` before seeing my changes re-appear magically. I have pushed my changes that contains the `test` that is `getWhenUsingCustomAuthenticationManagerResolverThenUsesItAccordingly` using `CustomAuthenticationManagerResolverConfig` in `OAuth2ResourceServerSpecTests` and when running the test I get the follwoing error: 

```
java.lang.AssertionError: 
Expected: a string starting with "Bearer error=\"mock-failure\""
     but: was "Bearer"

> GET 
> WebTestClient-Request-Id: [1]
> Authorization: [Bearer eyJhbGciOiJSUzI1NiJ9.eyJzdWIiOiJtb2NrLXN1YmplY3QiLCJzY29wZSI6Im1lc3NhZ2U6cmVhZCIsImV4cCI6NDY4ODY0MTQxM30.cRl1bv_dDYcAN5U4NlIVKj8uu4mLMwjABF93P4dShiq-GQ-owzaqTSlB4YarNFgV3PKQvT9wxN1jBpGribvISljakoC0E8wDV-saDi8WxN-qvImYsn1zLzYFiZXCfRIxCmonJpydeiAPRxMTPtwnYDS9Ib0T_iA80TBGd-INhyxUUfrwRW5sqKRbjUciRJhpp7fW2ZYXmi9iPt3HDjRQA4IloJZ7f4-spt5Q9wl5HcQTv1t4XrX4eqhVbE5cCoIkFQnKPOc-jhVM44_eazLU6Xk-CCXP8C_UT5pX0luRS2cJrVFfHp2IR_AWxC-shItg6LNEmNFD4Zc-JLZcr0Q86Q]

No content

< 401 UNAUTHORIZED Unauthorized
< Pragma: [no-cache]
< Expires: [0]
< X-Frame-Options: [DENY]
< X-Content-Type-Options: [nosniff]
< Referrer-Policy: [no-referrer]
< WWW-Authenticate: [Bearer]
< Cache-Control: [no-cache, no-store, max-age=0, must-revalidate]
< X-XSS-Protection: [1 ; mode=block]

0 bytes of content (unknown content-type).

	at org.springframework.test.web.reactive.server.ExchangeResult.assertWithDiagnostics(ExchangeResult.java:200)
	at org.springframework.test.web.reactive.server.HeaderAssertions.value(HeaderAssertions.java:81)
	at org.springframework.security.config.web.server.OAuth2ResourceServerSpecTests.getWhenUsingCustomAuthenticationManagerResolverThenUsesItAccordingly(OAuth2ResourceServerSpecTests.java:252)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.springframework.test.context.junit4.statements.RunBeforeTestExecutionCallbacks.evaluate(RunBeforeTestExecutionCallbacks.java:74)
	at org.springframework.test.context.junit4.statements.RunAfterTestExecutionCallbacks.evaluate(RunAfterTestExecutionCallbacks.java:84)
	at org.springframework.test.context.junit4.statements.RunBeforeTestMethodCallbacks.evaluate(RunBeforeTestMethodCallbacks.java:75)
	at org.springframework.test.context.junit4.statements.RunAfterTestMethodCallbacks.evaluate(RunAfterTestMethodCallbacks.java:86)
	at org.springframework.security.config.test.SpringTestRule$1.evaluate(SpringTestRule.java:36)
	at org.springframework.test.context.junit4.statements.SpringRepeat.evaluate(SpringRepeat.java:84)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.runChild(SpringJUnit4ClassRunner.java:251)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.runChild(SpringJUnit4ClassRunner.java:97)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.springframework.test.context.junit4.statements.RunBeforeTestClassCallbacks.evaluate(RunBeforeTestClassCallbacks.java:61)
	at org.springframework.test.context.junit4.statements.RunAfterTestClassCallbacks.evaluate(RunAfterTestClassCallbacks.java:70)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.run(SpringJUnit4ClassRunner.java:190)
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:86)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:38)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:538)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:760)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:460)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:206)
Caused by: java.lang.AssertionError: 
Expected: a string starting with "Bearer error=\"mock-failure\""
     but: was "Bearer"
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:8)
	at org.springframework.test.web.reactive.server.HeaderAssertions.lambda$value$1(HeaderAssertions.java:81)
	at org.springframework.test.web.reactive.server.ExchangeResult.assertWithDiagnostics(ExchangeResult.java:197)
	... 34 more
```
even though I have stubbed the authentication manager to return `error` on authentication which leads me to think that the `mocked authentication manager` is not used 🤔

jzheaux commented:
@rhamedy, I think the problem is that when just `authenticationManagerResolver()` is called, no filter is getting added.

I'd recommend something similar to this:

```java
if (this.authenticationManagerResolver != null) {
    AuthenticationWebFilter oauth2 = new AuthenticationWebFilter(this.authenticationManagerResolver);
    oauth2.setServerAuthenticationConverter(bearerTokenConverter);
    oauth2.setAuthenticationFailureHandler(new 
            ServerAuthenticationEntryPointFailureHandler(entryPoint));
    http.addFilterAt(oauth2, SecurityWebFiltersOrder.AUTHENTICATION);
} else if (this.jwt != null) {
    this.jwt.configure(http);
} else if (this.opaqueToken != null) {
    this.jwt.configure(http);
}
```

The extra support in `jwt.configure` and `opaqueToken.configure` for resolvers is unnecessary, which aligns with the fact that these are mutually exclusive configurations.

rhamedy commented:
Hi @jzheaux thanks for hint above, that seemed to be the issue with failing tests. I squashed my commits and pushed my changes. Please let me know if there is need for more tests and/or javadoc.

jzheaux commented on web/src/main/java/org/springframework/security/web/server/authentication/AuthenticationWebFilter.java:
```diff
@@ -51,18 +53,23 @@
 public class AuthenticationWebFilter implements WebFilter {
-
-	private final ReactiveAuthenticationManager authenticationManager;
+	private ReactiveAuthenticationManagerResolver<ServerHttpRequest> authenticationManagerResolver;
```
Is there any reason this can't be `final`? Let's make it `final` if it can be.

rhamedy replied:
I don't see a reason why not - `final` it is 👍

jzheaux commented:
Thanks, @rhamedy, this is now merged into `master`!
